In [51]:
# Import pandas packages
import pandas as pd

In [52]:
# Read the csv file
salaries_data = pd.read_csv("Salaries.csv")

C:\Users\Black Concept\AppData\Local\Temp\ipykernel_14084\775167987.py:2: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  salaries_data = pd.read_csv("Salaries.csv")


In [53]:
# Get a general overview if the imported dataset
salaries_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148049 non-null  object 
 4   OvertimePay       148654 non-null  object 
 5   OtherPay          148654 non-null  object 
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(3), int64(2), object(8)
memory usage: 14.7+ MB


In [54]:
# Check the top 5 rows of the dataset
salaries_data.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

### DATA CLEANING

Before analysing our data, let us take a look at a general overview of the data and see if there are null values, wrong data types etc

In [55]:
salaries_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148049 non-null  object 
 4   OvertimePay       148654 non-null  object 
 5   OtherPay          148654 non-null  object 
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(3), int64(2), object(8)
memory usage: 14.7+ MB


The dataset contains 13 columns and 148654 entries for the columns. 

Looking at some of the columns such as Benefits, BasePay, Notes, and Status, we can already see that we have missing data as they are not up to 148654 entries.

Also, some of the columns have mismatched data types, a good example would be BasePay, OvertimePay, OtherPay, Benefits which are supposed to be floats so we can carry out quantitative analysis on them but they are in object form.

Lastly, we have the notes column which is not needed and that has even been made evident by the fact that it has zero entry.

In [56]:
# Let us break the dataset into two categories- the columns with properly matched data types and the one that is mismatched

# Subsetting the columns with proper data types
proper_df = salaries_data[["Id", "EmployeeName", "JobTitle", "TotalPay", "TotalPayBenefits", "Year", "Agency", "Status"]]

In [57]:
# Viewing the proper dataset
proper_df.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

    TotalPay  TotalPayBenefits  Year         Agency Status  
0  567595.43         567595.43  2011  San Francisco    NaN  
1  538909.28         538909.28  2011  San Francisco    NaN  
2  335279.91         335279.91  2011  San Francisco    NaN  
3  332343.61         332343.61  2011  San Francisco    NaN  
4  326373.19         326373.19  2011  San Francisco    NaN

In [58]:
# Subsetting the columns with improper data types
improper_df = salaries_data[["BasePay", "OvertimePay", "OtherPay", "Benefits", "Notes"]]

In [59]:
# Viewing the improper dataset
improper_df.head()

BasePay OvertimePay   OtherPay Benefits  Notes
0  167411.18         0.0  400184.25      NaN    NaN
1  155966.02   245131.88  137811.38      NaN    NaN
2  212739.13   106088.18    16452.6      NaN    NaN
3    77916.0    56120.71   198306.9      NaN    NaN
4   134401.6      9737.0  182234.59      NaN    NaN

In [60]:
# Correcting the improper columns from object data types to numeric data types
improper_df_corrected = improper_df.apply(pd.to_numeric, errors="coerce")

In [61]:
# Let us take a look at the corrected improper columns to see their data types
improper_df_corrected.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   BasePay      148045 non-null  float64
 1   OvertimePay  148650 non-null  float64
 2   OtherPay     148650 non-null  float64
 3   Benefits     112491 non-null  float64
 4   Notes        0 non-null       float64
dtypes: float64(5)
memory usage: 5.7 MB


Now that our columns with mismatched data types have been corrected, it is time to join the dataset back together and continue our data cleaning exercise

In [62]:
# Concatenate the two datasets together
proper_salaries_data = pd.concat([proper_df, improper_df_corrected], axis=1)

In [63]:
# Let us take a look at our dataset
proper_salaries_data.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

    TotalPay  TotalPayBenefits  Year         Agency Status    BasePay  \
0  567595.43         567595.43  2011  San Francisco    NaN  167411.18   
1  538909.28         538909.28  2011  San Francisco    NaN  155966.02   
2  335279.91         335279.91  2011  San Francisco    NaN  212739.13   
3  332343.61         332343.61  2011  San Francisco    NaN   77916.00   
4  326373.19         326373.19  2011  San Francisco    NaN  134401.60   

   OvertimePay   OtherPay  Benefits  Notes  
0         0.00  400184.25       NaN    NaN  
1    245131.88  137811.38       NaN    NaN  
2    106088.18   16452.60       NaN    NaN  
3     56120.71  198306.90       NaN    NaN  
4      9737.00  182234.59       NaN    NaN

In [64]:
# From the info below, we now have the data types correctly matched to the columns
proper_salaries_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   TotalPay          148654 non-null  float64
 4   TotalPayBenefits  148654 non-null  float64
 5   Year              148654 non-null  int64  
 6   Agency            148654 non-null  object 
 7   Status            38119 non-null   object 
 8   BasePay           148045 non-null  float64
 9   OvertimePay       148650 non-null  float64
 10  OtherPay          148650 non-null  float64
 11  Benefits          112491 non-null  float64
 12  Notes             0 non-null       float64
dtypes: float64(7), int64(2), object(4)
memory usage: 14.7+ MB


In [65]:
# Shape of the proper dataset
shape = proper_salaries_data.shape
print("Number of rows:",shape[0])
print("Number of columns:",shape[1])

Number of rows: 148654
Number of columns: 13


In [66]:
# Let us handle missing data

# Check null data for each column
proper_salaries_data.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
BasePay                609
OvertimePay              4
OtherPay                 4
Benefits             36163
Notes               148654
dtype: int64

In [67]:
# Let us put the null values in percentages and qauntify it relatively
def percentage_null(data):
    missing_percentages = {}
    total_rows = len(data)
    for column in data:
        missing_count = data[column].isnull().sum()
        missing_percent = (missing_count / total_rows) * 100
        missing_percentages[column] = missing_percent
    return missing_percentages

In [68]:
percentage_null(proper_salaries_data)

{'Id': 0.0,
 'EmployeeName': 0.0,
 'JobTitle': 0.0,
 'TotalPay': 0.0,
 'TotalPayBenefits': 0.0,
 'Year': 0.0,
 'Agency': 0.0,
 'Status': 74.35723223054879,
 'BasePay': 0.40967616074912216,
 'OvertimePay': 0.002690812221669111,
 'OtherPay': 0.002690812221669111,
 'Benefits': 24.326960593055013,
 'Notes': 100.0}

The following columns would be dropped and the reasons are stated below as well:
1. ID: This is the employee ID which is not exactly needed for the type of questions we aim to answer with this data and besides, the index values can substitute for it
2. Notes: Notes has zero entries, it does not have any data to be analysed
3. Agency: The agency column has just one single entry and that is San Francisco. The data is for San Francisco, there is no need repeating that in the dataset
4. Status: Out of 148654 entries that we are supposed to have for status column, over 70% of the data is missing and that is a very high number. Whatever data filling method we use, It would not represent the true state of things so its better to drop it.

In [69]:
# Drop the ID, Notes, Agency, and Status columns and assign the datafram to a new variable since I am not dropping inplace
proper_salaries_data2 = proper_salaries_data.drop(["Id", "Notes", "Agency", "Status"], axis=1)

In [70]:
proper_salaries_data2

EmployeeName                                        JobTitle  \
0          NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1            GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2          ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3       CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4         PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
...                   ...                                             ...   
148649      Roy I Tillery                                       Custodian   
148650       Not provided                                    Not provided   
148651       Not provided                                    Not provided   
148652       Not provided                                    Not provided   
148653          Joe Lopez                      Counselor, Log Cabin Ranch   

         TotalPay  TotalPayBenefits  Year    BasePay  OvertimePay   OtherPay  \
0       567595.43         567595.43  2011  167411.18         0.00  400184.25   
1       538909.28         538909.28  2011  155966.02    245131.88  137811.38   
2       335279.91         335279.91  2011  212739.13    106088.18   16452.60   
3       332343.61         332343.61  2011   77916.00     56120.71  198306.90   
4       326373.19         326373.19  2011  134401.60      9737.00  182234.59   
...           ...               ...   ...        ...          ...        ...   
148649       0.00              0.00  2014       0.00         0.00       0.00   
148650       0.00              0.00  2014        NaN          NaN        NaN   
148651       0.00              0.00  2014        NaN          NaN        NaN   
148652       0.00              0.00  2014        NaN          NaN        NaN   
148653    -618.13           -618.13  2014       0.00         0.00    -618.13   

        Benefits  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
148649       0.0  
148650       NaN  
148651       NaN  
148652       NaN  
148653       0.0  

[148654 rows x 9 columns]

In [71]:
proper_salaries_data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   EmployeeName      148654 non-null  object 
 1   JobTitle          148654 non-null  object 
 2   TotalPay          148654 non-null  float64
 3   TotalPayBenefits  148654 non-null  float64
 4   Year              148654 non-null  int64  
 5   BasePay           148045 non-null  float64
 6   OvertimePay       148650 non-null  float64
 7   OtherPay          148650 non-null  float64
 8   Benefits          112491 non-null  float64
dtypes: float64(6), int64(1), object(2)
memory usage: 10.2+ MB


In [72]:
# We have a couple of rows where EmployeeName, JobTitle, and Payment details were not provided, we should drop them
proper_salaries_data2[(proper_salaries_data2["EmployeeName"] == "Not provided") & 
                      (proper_salaries_data2["JobTitle"] == "Not provided")]

EmployeeName      JobTitle  TotalPay  TotalPayBenefits  Year  BasePay  \
148646  Not provided  Not provided       0.0               0.0  2014      NaN   
148650  Not provided  Not provided       0.0               0.0  2014      NaN   
148651  Not provided  Not provided       0.0               0.0  2014      NaN   
148652  Not provided  Not provided       0.0               0.0  2014      NaN   

        OvertimePay  OtherPay  Benefits  
148646          NaN       NaN       NaN  
148650          NaN       NaN       NaN  
148651          NaN       NaN       NaN  
148652          NaN       NaN       NaN

In [73]:
proper_salaries_data2.drop(proper_salaries_data2[(proper_salaries_data2["EmployeeName"] == "Not provided") & 
                                                 (proper_salaries_data2["JobTitle"] == "Not provided")].index, inplace=True)

In [74]:
proper_salaries_data2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 148650 entries, 0 to 148653
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   EmployeeName      148650 non-null  object 
 1   JobTitle          148650 non-null  object 
 2   TotalPay          148650 non-null  float64
 3   TotalPayBenefits  148650 non-null  float64
 4   Year              148650 non-null  int64  
 5   BasePay           148045 non-null  float64
 6   OvertimePay       148650 non-null  float64
 7   OtherPay          148650 non-null  float64
 8   Benefits          112491 non-null  float64
dtypes: float64(6), int64(1), object(2)
memory usage: 11.3+ MB


In [75]:
# Shockingly, we still have occurrencies of missing data from BasePay, OvertimePay, OtherPay, and Benefits
# They are numerical values and we could fill the missing values with the median values

median_BasePay = proper_salaries_data2["BasePay"].median()
median_OvertimePay = proper_salaries_data2["OvertimePay"].median()
median_OtherPay = proper_salaries_data2["OtherPay"].median()
median_Benefits = proper_salaries_data2["Benefits"].median()

proper_salaries_data2["BasePay"] = proper_salaries_data2["BasePay"].fillna(median_BasePay)
proper_salaries_data2["OvertimePay"] = proper_salaries_data2["OvertimePay"].fillna(median_OvertimePay)
proper_salaries_data2["OtherPay"] = proper_salaries_data2["OtherPay"].fillna(median_OtherPay)
proper_salaries_data2["Benefits"] = proper_salaries_data2["Benefits"].fillna(median_Benefits)

In [76]:
proper_salaries_data2.info()
# Finally, we have a clean data

<class 'pandas.core.frame.DataFrame'>
Int64Index: 148650 entries, 0 to 148653
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   EmployeeName      148650 non-null  object 
 1   JobTitle          148650 non-null  object 
 2   TotalPay          148650 non-null  float64
 3   TotalPayBenefits  148650 non-null  float64
 4   Year              148650 non-null  int64  
 5   BasePay           148650 non-null  float64
 6   OvertimePay       148650 non-null  float64
 7   OtherPay          148650 non-null  float64
 8   Benefits          148650 non-null  float64
dtypes: float64(6), int64(1), object(2)
memory usage: 11.3+ MB


### EXPLORATORY DATA ANALYSIS

In [77]:
# Find occurence of the employee names (Top 5)
proper_salaries_data2["EmployeeName"].value_counts(sort=True).head()

Kevin Lee       13
William Wong    11
Richard Lee     11
Steven Lee      11
KEVIN LEE        9
Name: EmployeeName, dtype: int64

In [78]:
# Find the unique job titles
proper_salaries_data2["JobTitle"].unique()

array(['GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY',
       'CAPTAIN III (POLICE DEPARTMENT)',
       'WIRE ROPE CABLE MAINTENANCE MECHANIC', ...,
       'Forensic Toxicologist Supervis', 'Conversion', 'Cashier 3'],
      dtype=object)

In [79]:
# Find the number of unique job titles
proper_salaries_data2["JobTitle"].nunique()

2158

In [105]:
# Total number of jobs that contain "Captain"
proper_salaries_data2["JobTitle"].str.contains("CAPTAIN", case=False).sum()

552

In [81]:
# Display top 5 employee names from fire department
proper_salaries_data2[proper_salaries_data2["JobTitle"].str.contains("fire",case=False)]["EmployeeName"].head()

4        PATRICK GARDNER
6              ALSON LEE
8         MICHAEL MORRIS
9     JOANNE HAYES-WHITE
10         ARTHUR KENNEY
Name: EmployeeName, dtype: object

In [82]:
# Maximum, Minimum, and Average Base Pay
proper_salaries_data2["BasePay"].describe()

count    148650.000000
mean      66320.084634
std       42677.603433
min        -166.010000
25%       33846.057500
50%       65007.450000
75%       94609.715000
max      319275.010000
Name: BasePay, dtype: float64

In [84]:
# Find the job title of Albert Pardini
proper_salaries_data2[proper_salaries_data2["EmployeeName"] == "ALBERT PARDINI"]["JobTitle"]

2    CAPTAIN III (POLICE DEPARTMENT)
Name: JobTitle, dtype: object

In [85]:
# How much does he make, including benefits
proper_salaries_data2[proper_salaries_data2["EmployeeName"] == "ALBERT PARDINI"]["TotalPayBenefits"]

2    335279.91
Name: TotalPayBenefits, dtype: float64

In [86]:
# Highest basepay
proper_salaries_data2["BasePay"].max()

319275.01

In [87]:
# Index of employee with the highest basepay
max_basepay = proper_salaries_data2["BasePay"].idxmax()
max_basepay

72925

In [88]:
# Employee with the highest basepay
proper_salaries_data2[proper_salaries_data2["BasePay"].max() == proper_salaries_data2["BasePay"]]["EmployeeName"]

72925    Gregory P Suhr
Name: EmployeeName, dtype: object

In [89]:
# Average BasePay of all employee by year
proper_salaries_data2.groupby("Year")["BasePay"].mean()

Year
2011    63595.956517
2012    65436.406857
2013    69555.662801
2014    66564.421924
Name: BasePay, dtype: float64

In [90]:
# Average BasePay of all employee by per job title
proper_salaries_data2.groupby("JobTitle")["BasePay"].mean()

JobTitle
ACCOUNT CLERK                                     43300.806506
ACCOUNTANT                                        46643.172000
ACCOUNTANT INTERN                                 28732.663958
ACPO,JuvP, Juv Prob (SFERS)                       62290.780000
ACUPUNCTURIST                                     66374.400000
                                                      ...     
X-RAY LABORATORY AIDE                             47664.773077
X-Ray Laboratory Aide                             46086.387100
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.910000
Youth Comm Advisor                                39077.957500
ZOO CURATOR                                       43148.000000
Name: BasePay, Length: 2158, dtype: float64

In [91]:
# Average BasePay of all employee with job title accountant
proper_salaries_data2[proper_salaries_data2["JobTitle"] == "ACCOUNTANT"]["BasePay"].mean()

46643.172

In [95]:
# Top 5 most common jobs
proper_salaries_data2["JobTitle"].value_counts(sort=True).head()

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

In [100]:
# Percentage of each Job title
proper_salaries_data2["JobTitle"].value_counts(sort=True, normalize=True) * 100

Transit Operator                        4.733266
Special Nurse                           2.952573
Registered Nurse                        2.513286
Public Svc Aide-Public Works            1.693912
Police Officer 3                        1.628658
                                          ...   
BUILDING SERVICES TECHNICIAN            0.000673
TRIAL DELAY REDUCTION COORDINATOR       0.000673
POWERHOUSE ELECTRICIAN SUPERVISOR II    0.000673
CHIEF NURSERY SPECIALIST                0.000673
Cashier 3                               0.000673
Name: JobTitle, Length: 2158, dtype: float64

In [94]:
# Export cleaned data
proper_salaries_data2.to_csv("Salaries Data.csv")